### DISCUSSION TITLE
ClassMap handle the value of an optional column - null or empty value

### DISCUSSION THREAD

**Discussion by vernetk** [JoshClose/CsvHelper#1871] [Q&A] [answered]
Hi everyone,

Version CsvHelper ==> 19.0.0.

I would like the value of my optional column to be null if the column is not present and equal to its value if the column is present.

Example with column Engine:
CDE;Engine
201901182;F2BFA602D*M ==> value Engine = "F2BFA602D*M"
201900579; ==> value Engine = ""
201900641;F2BFA602D*M ==> value Engine = "F2BFA602D*M"

Example without column Engine:
CDE
201901182 ==> value Engine = null
201900579 ==> value Engine = null

If the value is null then I do not update my data in BDD on the other hand if the value is empty then I update my data.

I tried a lot of things but finally I can do the opposite with the following code.
`Map(m => m.EngineName).Name(Translations.Resources.Translation.Import_Engine_Name).Optional().TypeConverterOption.NullValues(String.Empty);`

Thank you for your help.

**Comment by JoshClose** [comment]
So in the first example you want a missing field to be an empty string, and in the second example you want it to be null?

If that's the case, just using `Optional()` should do that.

**Comment by vernetk** [comment]
Thanks for you help but optional don't work for me. 

I have two case. 
In my first example, the column "Entity" is present and all the field is present but someone are empty. I Will never have missing field in this example. 

In my second example, the column "Entity" missing.

If I use only optional().
In my first example, value of empty field is empty ==> its OK
On my second example the column "Entity" missing. Value of missing field is empty ==> I don't want this, I want null value.

Can you confirm me what you told me previously ?  Maybe the problem is with my class. 

Becarefull my version of csvhelper is 19.0.0

**Comment by JoshClose** [comment, marked as answer]
```cs
void Main()
{
    var s = new StringBuilder();
    s.Append("Id,Name\r\n");
    s.Append("1,\r\n");
    var config = new CsvConfiguration(CultureInfo.InvariantCulture)
    {
    };
    using (var reader = new StringReader(s.ToString()))
    using (var csv = new CsvReader(reader, config))
    {
        csv.Configuration.RegisterClassMap<FooMap>();
        csv.GetRecords<Foo>().ToList().Dump();
    }
}

private class Foo
{
    public int Id { get; set; }
    public string Name { get; set; }
}

private class FooMap : ClassMap<Foo>
{
    public FooMap()
    {
        Map(m => m.Id);
        Map(m => m.Name).Optional();
    }
}
```
![image](https://user-images.githubusercontent.com/114971/135667022-0c65b6c9-0e41-444d-9dc1-49a7e4bdea32.png)
```cs
void Main()
{
    var s = new StringBuilder();
    s.Append("Id\r\n");
    s.Append("1\r\n");
    var config = new CsvConfiguration(CultureInfo.InvariantCulture)
    {
    };
    using (var reader = new StringReader(s.ToString()))
    using (var csv = new CsvReader(reader, config))
    {
        csv.Configuration.RegisterClassMap<FooMap>();
        csv.GetRecords<Foo>().ToList().Dump();
    }
}

private class Foo
{
    public int Id { get; set; }
    public string Name { get; set; }
}

private class FooMap : ClassMap<Foo>
{
    public FooMap()
    {
        Map(m => m.Id);
        Map(m => m.Name).Optional();
    }
}
```
![image](https://user-images.githubusercontent.com/114971/135667090-f17415b8-d76c-4a80-b1bb-31c7ba0ed60f.png)

**Reply by vernetk** [reply]
Thanks you for you time.
My problem came from my class.

**Comment by avdimdv** [comment]
Actually if you have header with 2 items and records with one

    s.Append("Id,Name\r\n");
    s.Append("1\r\n");

you will get exception

```
CsvHelper.MissingFieldException: 'Field at index '1' does not exist. You can ignore missing fields by setting MissingFieldFound to null.
IReader state:
   ColumnCount: 0
   CurrentIndex: 1
   HeaderRecord:
["Id","Name"]
IParser state:
   ByteCount: 0
   CharCount: 12
   Row: 2
   RawRow: 2
   Count: 1
   RawRecord:
1
```

The same happen when you have csv without header, that was the reason I found this thread

**Reply by JoshClose** [reply]
That is a missing field on that row. Optional means the field doesn't exist on any row, including the header. The column is optional, not the individual field on a row.

**Reply by avdimdv** [reply]
But if we have csv without header (the case I have) optional does not work throwing exception. 
I can workaround this but it seems that something is wrong here.

**Reply by JoshClose** [reply]
```cs
void Main()
{
    var s = new StringBuilder();
    s.Append("Id\r\n");
    s.Append("1\r\n");
    var config = new CsvConfiguration(CultureInfo.InvariantCulture)
    {
    };
    using (var reader = new StringReader(s.ToString()))
    using (var csv = new CsvReader(reader, config))
    {
        csv.Context.RegisterClassMap<FooMap>();
        csv.GetRecords<Foo>().ToList().Dump();
    }
}

private class Foo
{
    public int Id { get; set; }
    public string Name { get; set; }
}

private class FooMap : ClassMap<Foo>
{
    public FooMap()
    {
        Map(m => m.Id);
        Map(m => m.Name).Optional();
    }
}
```
![image](https://user-images.githubusercontent.com/114971/196487046-de91c0f8-580c-4384-812e-bae08be43ac0.png)

**Reply by avdimdv** [reply]
```
void Main()
{
	var s = new StringBuilder();
	s.Append("1\r\n");
	s.Append("2\r\n");
	var config = new CsvConfiguration(CultureInfo.InvariantCulture)
	{
		HasHeaderRecord = false
	};
	using (var reader = new StringReader(s.ToString()))
	using (var csv = new CsvReader(reader, config))
	{
		csv.Context.RegisterClassMap<FooMap>();
		csv.GetRecords<Foo>().ToList().Dump();
	}
}

```

**Reply by JoshClose** [reply]
That's expected. The XML docs for `Optional` state `Ignore the member when reading if no matching field name can be found.` If headers are off, there is no field name.
